In [1]:
#importing all the libraries

import os
import pandas as pd
from sklearn.feature_selection import SelectFromModel
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn import tree
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
import matplotlib.pyplot as plt 
import seaborn as sns
import numpy as np
from sklearn.model_selection import cross_val_score 
from sklearn.feature_selection import SelectKBest, mutual_info_classif, chi2, f_classif
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score , confusion_matrix, precision_score, recall_score, classification_report
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import GridSearchCV
from sklearn.tree import export_graphviz
import graphviz
from sklearn.model_selection import LeaveOneOut
import warnings
warnings.filterwarnings("ignore")

In [2]:
# Datasets Loading Process

# Dataset 1 : CTU-133 
# Dataset 2 : IDS2017
# Dataset 3 : CIC-IDS2018 
# Dataset 4 : UNB-CA Botnet

# In order to have a proper and clean code, i will implement 4 functions, each function when called will 
# automatically load a dataset, and you can type help( function_name ) to get a description on the desired Dataset
# as i will implement a docstring containing the description


# You must keep in mind that the Datasets i'm using are extremely huge ( up to 79 GB of data) so until i discover 
# a way to optimize it, computations will be taking time

In [3]:
def load_CTU_13():
    '''
CTU-13 is a publicly available dataset of botnet traffic generated in a controlled environment. 
It contains 13 captures of different botnet traffic scenarios, including a mix 
of normal traffic and botnet traffic.The dataset includes both network traffic data and 
malware binaries for analysis. CTU-13 is a valuable resource for researchers and practitioners 
working in the field of network security and provides a realistic simulation of botnet trafficthat
can be used to evaluate the effectiveness of different detection and prevention techniques.
    '''

In [3]:
def load_IDS2017():
    '''
    IDS2017 is a publicly available dataset of network traffic generated in a realistic environment 
    to evaluate the performance of intrusion detection systems (IDS). It contains 5 days of network traffic
    captured from a real-world network environment, including a mix of normal traffic and various types 
    of attacks, such as botnet, DDoS, and brute-force attacks. The dataset includes both raw network traffic 
    data and preprocessed flow data, making it suitable for various types of IDS research. IDS2017 provides 
    a valuable resource for researchers and practitioners to evaluate and compare the effectiveness of
    different IDS techniques in detecting and mitigating various types of attacks.
        '''
    # we will start by loading the first file of our second dataset just to test 
    
    # Load the CSV file into a pandas DataFrame object
    df_IDS2017 = pd.read_csv('../Datasets/IDS_2017/Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv')
    
    # Print the first 5 rows of the DataFrame
    print(df_IDS2017.head())
    

In [21]:
load_IDS2017()

(225745, 79)


In [4]:
# Quick test to concatenate all the ID2017 csvs into one dataframe 

# Set the directory where your CSV files are located
directory = '../Datasets/IDS_2017'

# Create an empty list to store your DataFrames
df_list = []

# Loop over all CSV files in the directory and read them into separate DataFrames
for filename in os.listdir(directory):
    if filename.endswith('.csv'):
        filepath = os.path.join(directory, filename)
        df = pd.read_csv(filepath)
        df_list.append(df)

# Concatenate all DataFrames into a single DataFrame
combined_df = pd.concat(df_list, ignore_index=True)

# Print the first 5 rows of the combined DataFrame
print(combined_df.shape)

(2830743, 79)


# Task Done With success 1 file's shape is  (225745, 79), while concatenated files are (2830743, 79)

In [4]:
def load_CIC_IDS2018 ():
    '''
CIC-IDS2018 is a publicly available dataset of network traffic generated in a realistic environment 
to evaluate the performance of intrusion detection systems (IDS). It contains 9 days of network traffic
captured from a real-world network environment, including a mix of normal traffic and various types of 
attacks, such as botnet, DDoS, and malware attacks. The dataset includes both raw network traffic data 
and preprocessed flow data, making it suitable for various types of IDS research. CIC-IDS2018 provides 
a valuable resource for researchers and practitioners to evaluate and compare the effectiveness of different 
IDS techniques in detecting and mitigating various types of attacks in a realistic and complex 
network environment.
    '''

In [5]:
def load_UNB_CA_Botnet():
    '''
UNB-CA Botnet is a publicly available dataset of network traffic generated in a controlled environment 
to evaluate the effectiveness of different detection and prevention techniques against botnet attacks. 
The dataset contains traffic captures from different botnet scenarios, including HTTP, DNS, and P2P botnets,
as well as normal traffic. The dataset includes both raw network traffic data and preprocessed flow data, 
making it suitable for various types of botnet research. UNB-CA Botnet provides a valuable resource 
for researchers and practitioners to develop and evaluate new botnet detection and prevention techniques
in a controlled environment, allowing for better understanding and mitigation of botnet attacks.
    '''